### PR TITLE
[#1818] Twitter Editor of awesomeness

### PR DESCRIPTION
**Pull request by mjschranz**
https://webmademovies.lighthouseapp.com/projects/65733/tickets/1818-twitter-editor

### CODE REVIEW

**Comment by ghost** [comment]
#### From: Bot.io (butyr.org)

---

# Received

Command **cmd_preview** from @k88hudson received. Current queue size: 0

Live output at: http://botio.butyr.org:8000/686356d5cedff9f/output.txt

**Comment by ghost** [comment]
#### From: Bot.io (butyr.org)

---

# Success

Full output at http://botio.butyr.org:8000/686356d5cedff9f/output.txt

Total script time: 0.85 mins

Server is running on http://pr778.butyr.org/templates/basic/
Run the Test Suite: http://pr778.butyr.org/test/

**Comment by ghost** [comment]
#### From: Bot.io (butyr.org)

---

# Received

Command **cmd_preview** from @k88hudson received. Current queue size: 0

Live output at: http://botio.butyr.org:8000/8a623139b74b98f/output.txt

**Comment by ghost** [comment]
#### From: Bot.io (butyr.org)

---

# Success

Full output at http://botio.butyr.org:8000/8a623139b74b98f/output.txt

Total script time: 0.57 mins

Server is running on http://pr778.butyr.org/templates/basic/
Run the Test Suite: http://pr778.butyr.org/test/

File: templates/assets/plugins/twitter/popcorn.twitter.js [hunk 1]
```diff
@@ -2,6 +2,17 @@
 
 (function ( Popcorn, global ) {
 
+  function normalize( value, min, max ) {
+    value = value | 0;
```
Comment (cadecairos) [inline templates/assets/plugins/twitter/popcorn.twitter.js]: one line this Fn:

```
return Math.max( Math.min( value || 0, max ), min );
```

File: templates/assets/editors/twitter/twitter-editor.js [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+ * obtain one at http://www.mozillapopcorn.org/butter-license.txt */
+
+( function( Butter ) {
+
+  Butter.Editor.register( "twitter", "load!{{baseDir}}templates/assets/editors/twitter/twitter-editor.html",
+    function( rootElement, butter, compiledLayout ) {
+
+    var _this = this;
+
+    var _rootElement = rootElement,
```
Comment (dseif) [inline templates/assets/editors/twitter/twitter-editor.js]: merge these two var blocks into one

File: templates/assets/editors/twitter/twitter-editor.js [hunk 3]
```diff
@@ -0,0 +1,187 @@
+/*global google*/
+/* This Source Code Form is subject to the terms of the MIT license
+ * If a copy of the MIT license was not distributed with this file, you can
+ * obtain one at http://www.mozillapopcorn.org/butter-license.txt */
+
+( function( Butter ) {
```
Comment (dseif) [inline templates/assets/editors/twitter/twitter-editor.js]: `( function` -> `(function`

File: templates/assets/editors/twitter/twitter-editor.js [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+     * Member: setErrorState
+     *
+     * Sets the error state of the editor, making an error message visible
+     *
+     * @param {String} message: Error message to display
+     */
+    function setErrorState( message ) {
+      if ( message ) {
+        _messageContainer.innerHTML = message;
+        _messageContainer.parentNode.style.height = _messageContainer.offsetHeight + "px";
```
Comment (dseif) [inline templates/assets/editors/twitter/twitter-editor.js]: store a reference to `_messageContainer.parentNode` and `_messageContainer.parentNode.style`

File: templates/assets/editors/twitter/twitter-editor.js [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+
+      var container = _rootElement.querySelector( ".editor-options" ),
+          pluginOptions = {},
+          ignoreKeys = [
+            "search",
+            "username"
+          ];
+
+      function callback( elementType, element, trackEvent, name ) {
+        pluginOptions[ name ] = { element: element, trackEvent: trackEvent, elementType: elementType };
```
Comment (dseif) [inline templates/assets/editors/twitter/twitter-editor.js]: multi-lined object please

File: templates/assets/editors/twitter/twitter-editor.js [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+            userText = document.getElementById( "user-text" );
+
+        // Move start and end to first elements in editor
+        container.insertBefore( start, sourceLabel );
+        container.insertBefore( end, sourceLabel );
+
+        // Disable the user search box immediately
+        userText.disabled = true;
+
+        function updateSearch( e ) {console.log(e);
```
Comment (dseif) [inline templates/assets/editors/twitter/twitter-editor.js]: remove console.log

File: templates/assets/editors/twitter/twitter-editor.js [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+
+        function updateUser( evt ) {
+          trackEvent.update({
+            username: evt.target.value,
+            search: ""
+          });
+        }
+
+        function searchHandler() {
+          userText.removeEventListener( "blur", updateUser );
```
Comment (dseif) [inline templates/assets/editors/twitter/twitter-editor.js]: add the flag for `useCapture` here please, as I don't think all browsers default it to false

File: templates/assets/editors/twitter/twitter-editor.js [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+          searchText.disabled = false;
+
+          searchText.addEventListener( "blur", updateSearch, false );
+        }
+
+        searchText.addEventListener( "blur", updateSearch, false );
+        searchButton.addEventListener( "click", searchHandler, false );
+
+        function userHandler() {
+          searchText.removeEventListener( "blur", updateSearch );
```
Comment (dseif) [inline templates/assets/editors/twitter/twitter-editor.js]: same here

File: templates/assets/editors/twitter/twitter-editor.js [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+          searchText.disabled = true;
+          userText.disabled = false;
+
+          userText.addEventListener( "blur", updateUser, false );
+        }
```
Comment (dseif) [inline templates/assets/editors/twitter/twitter-editor.js]: any way we can meld `userHandler` and `searchHandler` into one function. The logic seems pretty similar and you can get a reference the to element it was triggered on through the event that gets passed in.

File: templates/assets/editors/twitter/twitter-editor.js [hunk 10]
```diff
@@ -0,0 +1,187 @@
...
+
+        userButton.addEventListener( "click", userHandler, false );
+
+        for ( key in pluginOptions ) {
```
Comment (dseif) [inline templates/assets/editors/twitter/twitter-editor.js]: just to be safe we should use `hasOwnProperty` here

File: templates/assets/editors/twitter/twitter-editor.js [hunk 11]
```diff
@@ -0,0 +1,187 @@
...
+              }
+            }
+          }
+        }
+      }
+
+      _this.createPropertiesFromManifest( trackEvent, callback, null, container, null, ignoreKeys );
+      attachHandlers();
+      _this.updatePropertiesFromManifest( trackEvent );
+
```
Comment (dseif) [inline templates/assets/editors/twitter/twitter-editor.js:173]: remove whitespace please

File: templates/assets/editors/twitter/twitter-editor.js [hunk 12]
```diff
@@ -0,0 +1,187 @@
...
+      open: function( parentElement, trackEvent ) {
+        _butter = butter;
+        // Update properties when TrackEvent is updated
+        trackEvent.listen( "trackeventupdated", function ( e ) {
+          _this.updatePropertiesFromManifest( e.target );
+          setErrorState( false );
+        });
+        setup( trackEvent );
+      },
+      close: function() {
```
Comment (dseif) [inline templates/assets/editors/twitter/twitter-editor.js:187]: we should probably unlisten trackeventupdated here so we don't have a ton of listeners left over when the user keeps opening/closing the editor

File: templates/assets/editors/twitter/twitter-editor.js [hunk 13]
```diff
@@ -0,0 +1,187 @@
...
+        _this.removeExtraHeadTags();
+      }
+    });
+
```
Comment (dseif) [inline templates/assets/editors/twitter/twitter-editor.js]: whitespace

File: templates/assets/editors/twitter/twitter-editor.js [hunk 14]
```diff
@@ -0,0 +1,187 @@
...
+  });
+
```
Comment (dseif) [inline templates/assets/editors/twitter/twitter-editor.js:194]: whitespace

File: templates/assets/editors/twitter/twitter.js [hunk 15]
```diff
@@ -0,0 +1,15 @@
+/*global EditorHelper*/
+
+EditorHelper.addPlugin( "twitter", function( trackEvent ) {
+  var _container,
+      media;
+
+  _container = trackEvent.popcornTrackEvent._container;
+  media = document.getElementById( trackEvent.track._media.target );
```
Comment (dseif) [inline templates/assets/editors/twitter/twitter.js]: these two declarations can be done on lines 4 and 5

File: templates/assets/editors/twitter/twitter.js [hunk 16]
```diff
@@ -0,0 +1,15 @@
...
+
+  if ( window.jQuery ) {
```
Comment (dseif) [inline templates/assets/editors/twitter/twitter.js]: we could probably add a 
     `if ( !window.jQuery ) {
       return;
     }`

at the top of this file so we can skip assigning variables and reduce indentation

File: templates/assets/plugins/twitter/popcorn.twitter.js [hunk 17]
```diff
@@ -11,6 +15,16 @@
...
+        start: {
+          elem: "input",
+          type: "number",
+          label: "Start"
+        },
+        end: {
+          elem: "input",
+          type: "number",
+          label: "End"
+        },
```
Comment (dseif) [inline templates/assets/plugins/twitter/popcorn.twitter.js]: maybe add `units: "seconds"` for start/end

File: templates/assets/plugins/twitter/popcorn.twitter.js [hunk 18]
```diff
@@ -22,7 +36,6 @@
           elem: "input",
           type: "text",
           label: "Tweets from User",
-          "default": "@popcornjs",
```
Comment (dseif) [inline templates/assets/plugins/twitter/popcorn.twitter.js:25]: why remove default?

File: templates/assets/plugins/twitter/popcorn.twitter.js [hunk 19]
```diff
@@ -40,24 +53,12 @@
...
-          transitionOut: "fly"
-          /*
-           * TODO when we sort out editor stuff
-           * Idea is the editors pass in a transitions object that tells us what transitions we want
-           * to apply.
-           */
+        transition: {
+          elem: "select",
+          options: [ "None", "Pop", "Fly Up", "Fly Down", "Sparkles" ],
+          values: [ "none", "pop", "flyUp", "flyDown", "sparkles" ],
```
Comment (dseif) [inline templates/assets/plugins/twitter/popcorn.twitter.js:61]: sparkles :D \0/

File: templates/assets/plugins/twitter/popcorn.twitter.js [hunk 20]
```diff
@@ -103,7 +95,9 @@
           imgLink,
           tweetTextCont,
           tweetUser,
-          tweetText;
+          tweetText,
+          top = normalize( options.top, 0, 97 ),
```
Comment (dseif) [inline templates/assets/plugins/twitter/popcorn.twitter.js]: why 97? maybe add a comment here as to why were using the magic number 97

File: templates/assets/plugins/twitter/popcorn.twitter.js [hunk 21]
```diff
@@ -111,6 +105,8 @@
 
       options._target = target;
 
+      options.numberOfTweets = options.numberOfTweets || 10;
```
Comment (dseif) [inline templates/assets/plugins/twitter/popcorn.twitter.js]: is this the default in the manifest for this property (10 that is)? If it is, then we should access it here so if it's changed in the manifest it gets updated here.

**Comment by ghost** [comment]
#### From: Bot.io (butyr.org)

---

# Received

Command **cmd_preview** from @dseif received. Current queue size: 0

Live output at: http://botio.butyr.org:8000/ef685c9386c7958/output.txt

**Comment by ghost** [comment]
#### From: Bot.io (butyr.org)

---

# Success

Full output at http://botio.butyr.org:8000/ef685c9386c7958/output.txt

Total script time: 0.62 mins

Server is running on http://pr778.butyr.org/templates/basic/
Run the Test Suite: http://pr778.butyr.org/test/

File: templates/assets/editors/twitter/twitter-editor.js [hunk 22]
```diff
@@ -0,0 +1,196 @@
...
+          setErrorState( false );
+        });
+        setup( trackEvent );
+      },
+      close: function() {
+        _this.removeExtraHeadTags();
+        trackEvent.unlisten( "trackeventupdated" );
+      }
+    });
+
```
Comment (dseif) [inline templates/assets/editors/twitter/twitter-editor.js]: whitespace here and a few others still

**Comment by ghost** [comment]
#### From: Bot.io (butyr.org)

---

# Received

Command **cmd_preview** from @dseif received. Current queue size: 1

Live output at: http://botio.butyr.org:8000/d73ee0b8fab76f9/output.txt

**Comment by ghost** [comment]
#### From: Bot.io (butyr.org)

---

# Received

Command **cmd_check** from @dseif received. Current queue size: 2

Live output at: http://botio.butyr.org:8000/5bcc51652b84e5e/output.txt

**Comment by ghost** [comment]
#### From: Bot.io (butyr.org)

---

# Success

Full output at http://botio.butyr.org:8000/d73ee0b8fab76f9/output.txt

Total script time: 0.73 mins

Server is running on http://pr778.butyr.org/templates/basic/
Run the Test Suite: http://pr778.butyr.org/test/

**Comment by ghost** [comment]
#### From: Bot.io (butyr.org)

---

# Failed

Full output at http://botio.butyr.org:8000/5bcc51652b84e5e/output.txt

Total script time: 1.08 mins

File: templates/assets/editors/twitter/twitter-editor.js [hunk 23]
```diff
@@ -0,0 +1,196 @@
...
+            searchText.disabled = true;
+            userText.disabled = false;
+
+            userText.addEventListener( "blur", updateUser, false );
+          }
+        }
+
+        searchText.addEventListener( "blur", updateSearch, false );
+        searchButton.addEventListener( "click", handler, false );
+        userButton.addEventListener( "click", handler, false );
```
Comment (dseif) [inline templates/assets/editors/twitter/twitter-editor.js:149]: there is no removeEventListener for this, will that be a problem?

Comment (mjschranz) [inline templates/assets/editors/twitter/twitter-editor.js:149]: No. These just simple toggle what elements are enabled and then toggle event listeners on the text boxes themselves.

**Comment by ghost** [comment]
#### From: Bot.io (butyr.org)

---

# Received

Command **cmd_preview** from @dseif received. Current queue size: 0

Live output at: http://botio.butyr.org:8000/b28c0501840d6be/output.txt

**Comment by ghost** [comment]
#### From: Bot.io (butyr.org)

---

# Failed

Full output at http://botio.butyr.org:8000/b28c0501840d6be/output.txt

Total script time: 0.89 mins

Server failed to start

**Comment by ghost** [comment]
#### From: Bot.io (butyr.org)

---

# Received

Command **cmd_preview** from @dseif received. Current queue size: 0

Live output at: http://botio.butyr.org:8000/361d080c3e2270a/output.txt

**Comment by ghost** [comment]
#### From: Bot.io (butyr.org)

---

# Failed

Full output at http://botio.butyr.org:8000/361d080c3e2270a/output.txt

Total script time: 0.99 mins

Server failed to start

**Comment by ghost** [comment]
#### From: Bot.io (butyr.org)

---

# Received

Command **cmd_preview** from @dseif received. Current queue size: 2

Live output at: http://botio.butyr.org:8000/cde59d7430f2be6/output.txt

**Comment by ghost** [comment]
#### From: Bot.io (butyr.org)

---

# Failed

Full output at http://botio.butyr.org:8000/cde59d7430f2be6/output.txt

Total script time: 0.77 mins

Server failed to start

**Comment by ghost** [comment]
#### From: Bot.io (butyr.org)

---

# Received

Command **cmd_preview** from @dseif received. Current queue size: 0

Live output at: http://botio.butyr.org:8000/07c4d09ce201b5e/output.txt

**Comment by ghost** [comment]
#### From: Bot.io (butyr.org)

---

# Failed

Full output at http://botio.butyr.org:8000/07c4d09ce201b5e/output.txt

Total script time: 0.57 mins

Server failed to start

**Comment by ghost** [comment]
#### From: Bot.io (butyr.org)

---

# Received

Command **cmd_preview** from @dseif received. Current queue size: 1

Live output at: http://botio.butyr.org:8000/2524ad8df45be6f/output.txt

**Comment by ghost** [comment]
#### From: Bot.io (butyr.org)

---

# Success

Full output at http://botio.butyr.org:8000/2524ad8df45be6f/output.txt

Total script time: 0.74 mins

Server is running on http://pr778.butyr.org/templates/basic/
Run the Test Suite: http://pr778.butyr.org/test/

**Comment by ghost** [comment]
#### From: Bot.io (butyr.org)

---

# Received

Command **cmd_preview** from @dseif received. Current queue size: 0

Live output at: http://botio.butyr.org:8000/0964ac7a5076b4c/output.txt

**Comment by ghost** [comment]
#### From: Bot.io (butyr.org)

---

# Success

Full output at http://botio.butyr.org:8000/0964ac7a5076b4c/output.txt

Total script time: 0.74 mins

Server is running on http://pr778.butyr.org/templates/basic/
Run the Test Suite: http://pr778.butyr.org/test/

**Comment by ghost** [comment]
#### From: Bot.io (butyr.org)

---

# Received

Command **cmd_preview** from @jbuck received. Current queue size: 0

Live output at: http://botio.butyr.org:8000/af48d3bddae790a/output.txt

**Comment by ghost** [comment]
#### From: Bot.io (butyr.org)

---

# Received

Command **cmd_check** from @jbuck received. Current queue size: 1

Live output at: http://botio.butyr.org:8000/cf535bddf40fe48/output.txt

**Comment by ghost** [comment]
#### From: Bot.io (butyr.org)

---

# Success

Full output at http://botio.butyr.org:8000/af48d3bddae790a/output.txt

Total script time: 0.70 mins

Server is running on http://pr778.butyr.org/templates/basic/
Run the Test Suite: http://pr778.butyr.org/test/

**Comment by ghost** [comment]
#### From: Bot.io (butyr.org)

---

# Failed

Full output at http://botio.butyr.org:8000/cf535bddf40fe48/output.txt

Total script time: 1.13 mins

File: templates/basic/index.html [hunk 24]
```diff
@@ -27,6 +27,8 @@
   <script src="../assets/editors/image/image.js" data-butter-exclude></script>
   <script src="../assets/editors/titles/titles.js" data-butter-exclude></script>
   <script src="../assets/editors/webpage/webpage.js" data-butter-exclude></script>
+  <script src="../assets/editors/googlemap.js" data-butter-exclude></script>
```
Comment (jbuck) [inline templates/basic/index.html:30]: Adding googlemap as well?

**Comment by ghost** [comment]
#### From: Bot.io (butyr.org)

---

# Received

Command **cmd_check** from @jbuck received. Current queue size: 0

Live output at: http://botio.butyr.org:8000/2e364a88a5cf67b/output.txt

**Comment by ghost** [comment]
#### From: Bot.io (butyr.org)

---

# Received

Command **cmd_preview** from @jbuck received. Current queue size: 1

Live output at: http://botio.butyr.org:8000/3399552f9df803e/output.txt

**Comment by ghost** [comment]
#### From: Bot.io (butyr.org)

---

# Success

Full output at http://botio.butyr.org:8000/2e364a88a5cf67b/output.txt

Total script time: 0.91 mins

**Comment by ghost** [comment]
#### From: Bot.io (butyr.org)

---

# Success

Full output at http://botio.butyr.org:8000/3399552f9df803e/output.txt

Total script time: 0.83 mins

Server is running on http://pr778.butyr.org/templates/basic/
Run the Test Suite: http://pr778.butyr.org/test/

**Pull request merged**